File: veggies_pro_Frontend/lib/models/product.dart
class UnitPrice {
  final String unit;
  final double step;
  final double baseQty;
  final double price;
  final double? compareAt;
  final double stock;

  UnitPrice({
    required this.unit,
    required this.step,
    required this.baseQty,
    required this.price,
    this.compareAt,
    required this.stock,
  });

  factory UnitPrice.fromJson(Map<String, dynamic> json) {
    return UnitPrice(
      unit: json['unit'],
      step: (json['step'] + 0.0),
      baseQty: (json['baseQty'] + 0.0),
      price: (json['price'] + 0.0),
      compareAt: json['compareAt']?.toDouble(),
      stock: (json['stock'] + 0.0),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'unit': unit,
      'step': step,
      'baseQty': baseQty,
      'price': price,
      'compareAt': compareAt,
      'stock': stock,
    };
  }

  double calculatePrice(double quantity) {
    return double.parse(((quantity / baseQty) * price).toStringAsFixed(2));
  }
}

class Category {
  final String id;
  final String name;
  final String? iconUrl;
  final int sort;
  final bool isActive;

  Category({
    required this.id,
    required this.name,
    this.iconUrl,
    required this.sort,
    required this.isActive,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['_id'] ?? json['id'],
      name: json['name'],
      iconUrl: json['iconUrl'],
      sort: json['sort'] ?? 0,
      isActive: json['isActive'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'iconUrl': iconUrl,
      'sort': sort,
      'isActive': isActive,
    };
  }
}

class Product {
  final String id;
  final String name;
  final String slug;
  final String categoryId;
  final List<String> images;
  final String? description;
  final List<UnitPrice> unitPrices;
  final double? rating;
  final bool isActive;

  Product({
    required this.id,
    required this.name,
    required this.slug,
    required this.categoryId,
    required this.images,
    this.description,
    required this.unitPrices,
    this.rating,
    required this.isActive,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['_id'] ?? json['id'],
      name: json['name'],
      slug: json['slug'],
      categoryId: json['categoryId']?['_id'] ?? json['categoryId'],
      images: List<String>.from(json['images'] ?? []),
      description: json['description'],
      unitPrices: (json['unitPrices'] as List<dynamic>)
          .map((unitPrice) => UnitPrice.fromJson(unitPrice))
          .toList(),
      rating: json['rating']?.toDouble(),
      isActive: json['isActive'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'slug': slug,
      'categoryId': categoryId,
      'images': images,
      'description': description,
      'unitPrices': unitPrices.map((unitPrice) => unitPrice.toJson()).toList(),
      'rating': rating,
      'isActive': isActive,
    };
  }

  String get firstImage => images.isNotEmpty ? images.first : '';
  
  double get minPrice {
    if (unitPrices.isEmpty) return 0.0;
    return unitPrices.map((up) => up.price).reduce((a, b) => a < b ? a : b);
  }
}

File: veggies_backend/src/models/Product.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

const unitPriceSchema = new Schema({
    unit: { type: String, enum: ['kg', 'g', 'pcs', 'bundle'], required: true },
    step: { type: Number, required: true },
    baseQty: { type: Number, required: true },
    price: { type: Number, required: true },
    compareAt: { type: Number },
    stock: { type: Number, required: true, min: 0 }
});

const productSchema = new Schema({
    name: { type: String, required: true, trim: true },
    slug: { type: String, required: true, unique: true },
    categoryId: { type: Schema.Types.ObjectId, ref: 'Category', required: true },
    images: [{ type: String }],
    description: { type: String },
    unitPrices: [unitPriceSchema],
    rating: { type: Number, min: 0, max: 5 },
    isActive: { type: Boolean, default: true },
    isDeal: { type: Boolean, default: false },
    dealEndsAt: { type: Date, default: null }
}, {
    timestamps: true
});

// Indexes for faster queries
productSchema.index({ categoryId: 1, isActive: 1 });
productSchema.index({ slug: 1 });
productSchema.index({ name: 'text', description: 'text' });

// Generate slug from name
productSchema.pre('save', function (next) {
    if (this.isModified('name') || !this.slug) {
        this.slug = this.name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');
    }
    next();
});

const Product = mongoose.model('Product', productSchema);

module.exports = { Product };
